src/passkey_child/passkey_child_common.c

```diff
@@ -319,3 +343,93 @@ register_key(struct passkey_data *data)
 
     return ret;
 }
+
+errno_t
+format_pem_public_key(TALLOC_CTX *mem_ctx, const struct passkey_data *data,
+                      char **_pem_key)
+{
+    BIO *bio_mem = NULL;
+    EVP_PKEY *evp_pkey = NULL;
+    unsigned char *public_key = NULL;
+    size_t pk_length;
+    char *pem_key;
+    long pem_size;
+    unsigned long err;
+    errno_t ret;
+
+    public_key = sss_base64_decode(mem_ctx, data->public_key, &pk_length);
+    if (public_key == NULL) {
+        DEBUG(SSSDBG_OP_FAILURE, "sss_base64_decode failed.\n");
+        ret = EINVAL;
+        goto done;
+    }
+
+    switch (data->type) {
+    case COSE_ES256:
+        ret = es256_pubkey_to_evp_pkey(mem_ctx, public_key, pk_length, &evp_pkey);
+        if (ret != EOK) {
+            goto done;
+        }
+        break;
+    case COSE_RS256:
+        ret = rs256_pubkey_to_evp_pkey(mem_ctx, public_key, pk_length, &evp_pkey);
+        if (ret != EOK) {
+            goto done;
+        }
+        break;
+    case COSE_EDDSA:
+        ret = eddsa_pubkey_to_evp_pkey(mem_ctx, public_key, pk_length, &evp_pkey);
+        if (ret != EOK) {
+            goto done;
+        }
+        break;
+    default:
+        DEBUG(SSSDBG_OP_FAILURE, "Invalid key type.\n");
+        ret = EINVAL;
+        goto done;
+        break;
+    }
+
+    bio_mem = BIO_new(BIO_s_mem());
+    if (bio_mem == NULL) {
+        DEBUG(SSSDBG_OP_FAILURE, "BIO_new failed.\n");
+        return ENOMEM;
```

Comment:
`goto done;`? Otherwise leaks keys.

Reply:
(I guess my comment to previous version triggered this.)

Reply:
yes, that makes sense. I didn't take into account that the keys have already been allocated before calling this function.